C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,314 @@
+	uint16_t TotalLength;
+	char ProviderName[ANYSIZE_ARRAY];
+};
+
+struct EventMetadata
+{
+	uint16_t TotalLength;
+	uint8_t Tag;
+	char EventName[ANYSIZE_ARRAY];
+};
+
+EVENT_DATA_DESCRIPTOR CreateProviderMetadata(const char* providerName)
+{
+	// Create packaged provider metadata structure.
+	// <https://learn.microsoft.com/en-us/windows/win32/etw/provider-traits>
+	const auto providerMetadataLength = (uint16_t)((strlen(providerName) + 1) + sizeof(uint16_t));
+	const auto providerMetadata = (struct detail::ProviderMetadata*)ExAllocatePoolWithTag(NonPagedPoolNx, providerMetadataLength, 'wteO');
+	if (providerMetadata == NULL)
+	{
+		return EVENT_DATA_DESCRIPTOR{};
+	}
+
+	RtlSecureZeroMemory(providerMetadata, providerMetadataLength);
+	providerMetadata->TotalLength = providerMetadataLength;
+	strcpy(providerMetadata->ProviderName, providerName);
+
+	// Create an EVENT_DATA_DESCRIPTOR pointing to the metadata.
+	EVENT_DATA_DESCRIPTOR providerMetadataDesc;
+	EventDataDescCreate(&providerMetadataDesc, providerMetadata, providerMetadata->TotalLength);
+	providerMetadataDesc.Type = EVENT_DATA_DESCRIPTOR_TYPE_PROVIDER_METADATA;  // Descriptor contains provider metadata.
+
+	return providerMetadataDesc;
+}
+
+size_t SizeOfFieldsMeta()
+{
+	return 0;
+}
+
+template<typename FieldValue, typename... Rest>
+size_t SizeOfFieldsMeta(const char* fieldName, int fieldType, FieldValue& fieldValue, Rest... rest)
+{
+	UNREFERENCED_PARAMETER(fieldType);
+	UNREFERENCED_PARAMETER(fieldValue);
+	return strlen(fieldName) + 1 + sizeof(uint8_t) + SizeOfFieldsMeta(rest...);
+}
+
+void SetFieldMetadata(uint8_t* current)
+{
+	UNREFERENCED_PARAMETER(current);
+}
+
+template<typename FieldValue, typename... Rest>
+void SetFieldMetadata(uint8_t* current, const char* fieldName, int fieldType, FieldValue& fieldValue, Rest... rest)
+{
+	UNREFERENCED_PARAMETER(fieldValue);
+
+	strcpy((char*)current, fieldName);
+	current[strlen(fieldName) + 1] = (uint8_t)fieldType;
+
+	current += strlen(fieldName) + 1 + sizeof(uint8_t);
+	SetFieldMetadata(current, rest...);
+}
+
+template<typename... Arguments>
+EVENT_DATA_DESCRIPTOR CreateEventMetadata(const char* eventName, Arguments... args)
+{
+	// Allocate the total size of the event metadata structure.
+	//
+	// This consists of, in order:
+	//
+	// * The total length of the structure
+	// * The event tag byte (currently always 0)
+	// * The name of the event
+	// * An array of field metadata structures, which are:
+	//     * the name of the field
+	//     * a single byte for the type.
+	const auto eventMetadataHeaderLength = strlen(eventName) + 1 + sizeof(uint16_t) + sizeof(uint8_t);
+	const auto eventMetadataLength = (uint16_t)(eventMetadataHeaderLength + SizeOfFieldsMeta(args...));
+	const auto eventMetadata = (struct detail::EventMetadata*)ExAllocatePoolWithTag(NonPagedPoolNx, eventMetadataLength, 'wteE');
+	if (eventMetadata == NULL)
+	{
+		return EVENT_DATA_DESCRIPTOR{};
+	}
+
+	RtlSecureZeroMemory(eventMetadata, eventMetadataLength);
```

Comment:
We can just use memset here, the contents of allocated pools on modern windows should be zero anyways and this data is not a security concern.

Reply:
Wasn't the contents not being zeroed-out the problem seen in [this](https://github.com/mandiant/STrace/pull/11) PR? [ExAllocatePool2](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-exallocatepool2) zeroes the memory, but the rest of the project doesn't use it. I could just use memset though yeah.

Reply:
Yes, you are right, I was thinking of:

> ExAllocatePool2 allocates is zero initialized

But still, yes memset please